---
title: "DO enrichment analysis"
output: html_document
date: "2023-02-22"
---

```{r}
library(RColorBrewer)
curdir <- "~/Labolatory/JST_Mouse_Human/Tabula/Integration"

# renv::restore()
setwd(curdir)
library(here)
set_here(curdir)
source("run_seurat_GO_analysis.R")
source("count_marker_gene_GO.R")
library(DOSE)
read_tsv("cell_ontology_class_manual.txt",
         show_col_types = FALSE) -> cell_type_label
```

統合がうまく行っているか？
```{r}
integration <- tibble(organ = c("Bladder", "Bone_Marrow", "Fat", "Heart", "Kidney",
                                      "Large_Intestine", "Liver", "Lung", "Muscle",
                                      "Pancrease", "Skin", "Spleen", "Thymus", "Trachea"),
                      integration = c("Bad", "Good", "Good", "Good", "Bad", "Bad", "Good",
                                      "Good", "Bad", "Good", "Bad", "Bad", "Bad", "Good"))
```
NCBI Entriz IDを取得する
```{r}
# ファイルから読み込む
"../../Entrez_gene_ID/230214_hs_entrez_gene_id.tsv" %>%
  read_tsv(show_col_types = FALSE) -> entrez
```
ヒトマーカー遺伝子を利用して、各細胞種マーカーに濃縮されたDOを調べた
```{r}
hs_marker %>%
  dplyr::select(organ, cluster) %>%
  distinct -> target_list
result_table <- c()
for(i in 1:dim(target_list)[1]){
  target_list[i, 1] %>%
    as.character() -> organ
  target_list[i, 2] %>%
    unlist() %>%
    as.character %>%
    as.numeric -> clst_id
  
  hs_marker %>%
    filter(organ == organ & cluster == clst_id) %>%
    dplyr::select(Gene_stable_ID) %>%
    inner_join(entrez, by = "Gene_stable_ID") %>%
    dplyr::select(NCBI_gene_formerly_Entrezgene_ID) %>%
    distinct %>%
    unlist() ->  gene
  
  enrichDO(gene, pAdjustMethod = "fdr", qvalueCutoff = 0.05) %>%
    as_tibble %>%
    mutate(organ = organ, cluster = clst_id) %>%
    bind_rows(result_table, .) -> result_table 
}
do_enrichment_result <- result_table

# FDRの分布
do_enrichment_result %>%
  mutate(log_fdr = -log(p.adjust), label = paste(ID, organ, cluster)) %>%
  ggplot(aes(x = log_fdr)) +
    geom_histogram(binwidth = 1)
# DOがいくつの細胞クラスターから見つかるか？
do_enrichment_result %>%
    dplyr::select(ID, organ, cluster) %>%
    distinct %>% 
    dplyr::count(ID) %>%
    ggplot(aes(x = n)) + geom_histogram(binwidth = 1)
```
DO enrichment解析と、細胞種マーカーかつヒト-マウス共通DOを結合させて、プロット
テスト版
```{r}
result_table %>%
  mutate(cluster = as.factor(cluster)) %>%
  dplyr::select(ID, organ, cluster) %>%
  inner_join(hs_mm_marker_disease, by = c("ID" = "DO_Disease_ID",
                                          "organ", "cluster")) %>%
  mutate(cell_ontology_label = paste(organ, customclassif, sep = "_")) %>%
  group_by(cell_ontology_label) %>%
  dplyr::count(share) %>%
  ungroup() %>%
  tidyr::complete(cell_ontology_label, share) %>%
  mutate(n = replace_na(n, 0)) %>%
  mutate(organ = str_split(cell_ontology_label,
                             pattern = "_", simplify = TRUE)[,1],
         ontology = str_split(cell_ontology_label,
                                      pattern = "_", simplify = TRUE)[,3],
         ontology_label = paste(str_split(cell_ontology_label,
                                      pattern = "_", simplify = TRUE)[,1],
                                str_split(cell_ontology_label,
                                      pattern = "_", simplify = TRUE)[,2],
                                sep = "_")) %>%
  inner_join(cell_type_label,
             by = c(ontology = "cell_ontology_class")) %>%
  ggplot(aes(x = reorder(summarized_class, n), y = n, fill = share)) + 
    geom_bar(stat= "identity", position = "dodge") +
    facet_grid(organ ~ large_summarized_class,
                     scale = "free_x", space = "free_x",
                     labeller = label_wrap_gen(5)) +
    theme_bw() +
    theme(axis.text.x =
              element_text(angle = 90, hjust = 1, vjust = 0.5),
          legend.position = "bottom") +
    NoLegend()
```
MGIのDisease ontologyと遺伝子の表をプロット
```{r}
plot_DO_enrichment <- function(x){
  result_table %>%
      mutate(cluster = as.factor(cluster)) %>%
      dplyr::select(ID, organ, cluster) %>%
      inner_join(x, by = c("ID" = "DO_Disease_ID",
                           "organ", "cluster"),
                 multiple = "all") %>%
      mutate(cell_ontology_label = paste(organ, customclassif, sep = "_")) %>%
      group_by(cell_ontology_label) %>%
      dplyr::count(share) %>%
      ungroup() %>%
      tidyr::complete(cell_ontology_label, share) %>%
      mutate(n = replace_na(n, 0)) %>%
      mutate(cell_ontology_label = gsub("Bone_Marrow", "Bone Marrow", 
                                        cell_ontology_label)) %>%
      mutate(cell_ontology_label = gsub("Large_Intestine", "Large Intestine",
                                        cell_ontology_label))%>%
      mutate(organ = str_split(cell_ontology_label,
                                 pattern = "_", simplify = TRUE)[,1],
             ontology = str_split(cell_ontology_label,
                                          pattern = "_", simplify = TRUE)[,3],
             ontology_label = paste(str_split(cell_ontology_label,
                                          pattern = "_", simplify = TRUE)[,1],
                                    str_split(cell_ontology_label,
                                          pattern = "_", simplify = TRUE)[,2],
                                    sep = "_")) %>%
      inner_join(cell_type_label,
                 by = c(ontology = "cell_ontology_class")) %>%
      ggplot(aes(x = reorder(ontology, n), y = n, fill = share)) + 
        geom_bar(stat= "identity", position = "dodge") +
        facet_grid(organ ~ large_summarized_class,
                         scale = "free_x", space = "free_x",
                         labeller = label_wrap_gen(5)) +
        theme_bw() +
        theme(axis.text.x =
                  element_text(angle = 90, hjust = 1, vjust = 0.5),
              legend.position = "bottom") 
}
# ヒト-マウス共通
plot_DO_enrichment(hs_mm_marker_disease) -> p1
# ヒト-マウス共通のDOだが、関連遺伝子が異なる
plot_DO_enrichment(hs_mm_marker_disease.different_bg) -> p2
# ヒトのみ
plot_DO_enrichment(hs_only_marker_disease) -> p3
```
マーカー遺伝子かつDO関連遺伝子を抽出したい
```{r}
# DO enrichment解析からIDと細胞種の情報を抽出した
do_enrichment_result %>%
  dplyr::select(ID, organ, cluster, geneID) -> outtable

# DO enrichment解析の結果から、遺伝子をtidyにした表を抽出する
output_table <- tibble()
convert_do_result_to_gene <- function(table_row_vector){
  table_row_vector %>%
    dplyr::select(geneID) %>%
    unlist %>%
    str_split("/", simplify = TRUE) %>%
    as.vector -> genes
  tibble(ID    = rep(unlist(table_row_vector[1, 1]), length(genes)),
         organ = rep(unlist(table_row_vector[1, 2]), length(genes)),
         cluster = rep(unlist(table_row_vector[1, 3]), length(genes)),
         geneID = genes)
}
for(i in 1:dim(outtable)[1]){
    print(i)
    convert_do_result_to_gene(outtable[i,]) %>%
        bind_rows(output_table, .,) -> output_table
}

# DO enrichment解析の結果はNCBI entrez gene IDなので、ENSEMBL IDに変換する
output_table %>%
    mutate(NCBI_gene_formerly_Entrezgene_ID = as.numeric(geneID)) %>%
    inner_join(entrez,
               by = c("NCBI_gene_formerly_Entrezgene_ID")) %>%
    dplyr::select(ID, organ, cluster, NCBI_gene_formerly_Entrezgene_ID,
                  Gene_stable_ID, Gene_name) %>%
    distinct -> all_do_gene_table

```
```{r}
# 細胞種ラベルを用意
hs_marker %>%
    dplyr::select(organ, customclassif) %>%
    distinct %>%
    mutate(cluster  = str_split(customclassif, pattern = "_", simplify = TRUE)[,1],
           ontology = str_split(customclassif, pattern = "_", simplify = TRUE)[,2]) %>%
    mutate(cell_label = paste(organ, cluster, sep = "_")) -> cell_type_ontology

# DO enrichment解析とマーカー遺伝子の結果を結合
all_do_gene_table %>%
  mutate(cluster = factor(cluster)) %>%
  inner_join(hs_marker, by = c("Gene_stable_ID", "Gene_name",
                               "organ", "cluster")) -> do_gene_table_marker
# DO enrichment解析とマーカー遺伝子の結果から、細胞種ラベルを変えてプロットする
do_gene_table_marker %>%
    dplyr::select(organ, cluster, Gene_name, share, Homology_type) %>%
    distinct %>%
    mutate(cell_label = paste(organ, cluster, sep = "_")) %>%
    group_by(cell_label) %>%
    dplyr::count(share) %>%
    ungroup() %>% 
    tidyr::complete(cell_label, share) %>%
    mutate(n = replace_na(n, 0)) %>%
    inner_join(cell_type_ontology, by = "cell_label") %>%
    inner_join(cell_type_label,
               by = c(ontology = "cell_ontology_class")) %>%
    mutate(organ = gsub("Bone_Marrow", "Bone Marrow", 
                                      organ)) %>%
    mutate(organ = gsub("Large_Intestine", "Large Intestine",
                                      organ))%>%
    ggplot(aes(x = reorder(ontology, n), y = n, fill = share)) + 
        geom_bar(stat= "identity", position = "dodge") +
        facet_grid(organ ~ large_summarized_class,
                   scale = "free_x", space = "free_x",
                   labeller = label_wrap_gen(5)) +
        theme_bw() +
        theme(axis.text.x =
                  element_text(angle = 90, hjust = 1, vjust = 0.5),
              legend.position = "bottom")
```
# あえてクラスター番号でプロット
```{r}
do_gene_table_marker %>%
    dplyr::select(organ, cluster, Gene_name, share, Homology_type) %>%
    distinct %>%
    mutate(cell_label = paste(organ, cluster, sep = "_")) %>%
    group_by(cell_label) %>%
    dplyr::count(share) %>%
    ungroup() %>% 
    tidyr::complete(cell_label, share) %>%
    mutate(n = replace_na(n, 0)) %>%
    inner_join(cell_type_ontology, by = "cell_label") %>%
    inner_join(cell_type_label,
               by = c(ontology = "cell_ontology_class")) %>%
    mutate(organ = gsub("Bone_Marrow", "Bone Marrow", 
                                      organ)) %>%
    mutate(organ = gsub("Large_Intestine", "Large Intestine",
                                      organ)) %>%
    mutate(cluster = factor(cluster, levels = 0:20)) %>%
    ggplot(aes(x = cluster, y = n, fill = share)) + 
        geom_bar(stat= "identity", position = "dodge") +
        facet_grid(organ ~ .,
                   scale = "free_x", space = "free_x",
                   labeller = label_wrap_gen(5)) +
        theme_bw() +
        theme(axis.text.x =
                  element_text(angle = 90, hjust = 1, vjust = 0.5),
              legend.position = "bottom")
```


細胞種ごとに、DOに関連した遺伝子のうち2種に共通するマーカー遺伝子をカウントした
```{r}
do_gene_table_marker %>%
    dplyr::select(ID, organ, cluster, Gene_name, share, Homology_type) %>%
    distinct %>%
    mutate(cell_label = paste(organ, cluster, sep = "_")) %>%
    group_by(cell_label, ID) %>%
    dplyr::count(share) %>%
    ungroup() %>% 
    tidyr::complete(cell_label, share) %>%
    mutate(organ = str_split(cell_label, pattern = "_", simplify = TRUE)[,1]) -> count_marker_gene_per_share

# DOはどの程度細胞種横断的に濃縮されているのか？
# 肝臓
count_marker_gene_per_share %>%
    filter(organ == "Liver") %>%
    ggplot(aes(x = ID, y = n, fill = share)) +
        geom_bar(stat ="identity", position = "fill") +
        facet_grid(cell_label ~ .) # 多くのDOが細胞種横断的に濃縮されている
```

細胞種ごとに、DOに関連した遺伝子のうち2種に共通するマーカー遺伝子カウントをプロットした
```{r}
hs_mm_marker_disease %>%
  dplyr::select(DO_Disease_ID) %>%
  distinct -> hs_mm_marker_disease.do

do_enrichment_result %>%
  mutate(cluster = factor(cluster)) %>%
  dplyr::select(ID, organ, cluster, p.adjust) %>%
  inner_join(do_gene_table_marker,
             by = c("ID", "organ", "cluster")) %>%
  mutate(cell_label = paste(organ, cluster, sep = "_")) %>%
  dplyr::select(cell_label, ID, p.adjust)  %>%
  distinct -> do_id_cell

count_marker_gene_per_share %>%
    filter(organ == "Liver") %>%
    inner_join(hs_mm_marker_disease.do, by = c("ID" = "DO_Disease_ID"))  %>%
    ungroup() %>%
    tidyr::complete(cell_label, organ, ID, share) %>%
    mutate(n = replace_na(n, 0)) %>%
    ggplot(aes(x = ID, y = n, fill = share)) +
    geom_bar(stat ="identity", position = "dodge") +
    facet_grid(cell_label ~ .) +
    theme_bw() +
    theme(axis.text.x =
              element_text(angle = 90, hjust = 1, vjust = 0.5),
          legend.position = "bottom")
```
細胞種に濃縮されたDOに関連する種特異的な細胞種マーカー遺伝子について、遺伝子の対応関係ごとにカウント
```{r}
do_gene_table_marker %>%
    dplyr::select(organ, cluster, Gene_name, share, Homology_type) %>%
    distinct %>%
    mutate(cell_label = paste(organ, cluster, sep = "_"),
           Homology_type = paste(share, Homology_type)) %>%
    group_by(cell_label) %>%
    dplyr::count(Homology_type) %>%
    ungroup() %>% 
    tidyr::complete(cell_label, Homology_type) %>%
    mutate(n = replace_na(n, 0)) %>%
    inner_join(cell_type_ontology, by = "cell_label") %>%
    inner_join(cell_type_label,
               by = c(ontology = "cell_ontology_class")) %>%
    mutate(organ = gsub("Bone_Marrow", "Bone Marrow", 
                                      organ)) %>%
    mutate(organ = gsub("Large_Intestine", "Large Intestine",
                                      organ)) %>%
    # filter(Homology_type == "Hs_only Hs_SS") %>% # ヒトSS遺伝子を抽出する場合
    ggplot(aes(x = reorder(ontology, n), y = n, fill = Homology_type)) + 
    geom_bar(stat= "identity", position = "dodge") +
    facet_grid(organ ~ large_summarized_class,
               scale = "free_x", space = "free_x",
               labeller = label_wrap_gen(5)) +
    theme_bw() +
    theme(axis.text.x =
              element_text(angle = 90, hjust = 1, vjust = 0.5),
          legend.position = "bottom")
```
細胞種に濃縮されたマーカー遺伝子に濃縮されたDOのうち、DO関連の細胞種マーカーにしめるヒト系統特異的な遺伝子の数
```{r}
do_gene_table_marker %>%
    dplyr::select(organ, cluster, Gene_name, share, Homology_type) %>%
    distinct %>%
    mutate(cell_label = paste(organ, cluster, sep = "_"),
           Homology_type = paste(share, Homology_type, sep = "_")) %>%
    group_by(cell_label) %>%
    dplyr::count(Homology_type) %>%
    ungroup() %>% 
    tidyr::complete(cell_label, Homology_type) %>%
    mutate(n = replace_na(n, 0)) %>%
    inner_join(cell_type_ontology, by = "cell_label") %>%
    inner_join(cell_type_label,
               by = c(ontology = "cell_ontology_class")) %>%
    mutate(organ = gsub("Bone_Marrow", "Bone Marrow", 
                        organ)) %>%
    mutate(organ = gsub("Large_Intestine", "Large Intestine",
                        organ)) %>%
    dplyr::select(organ, cluster, ontology, large_summarized_class, Homology_type, n) %>%
    pivot_wider(c(organ, cluster, ontology, large_summarized_class),
                names_from = Homology_type, values_from = n) %>%
    mutate(ratio = Hs_only_Hs_SS / sum(Hs_only_Ambigous, Hs_only_Hs_SS,
                                       Hs_only_ortholog_many2many,
                                       Hs_only_ortholog_one2many,
                                       Hs_only_ortholog_one2one,
                                       shared_Ambigous, shared_ortholog_many2many)) %>%
    ggplot(aes(x = reorder(ontology, ratio), y = ratio)) + 
    geom_bar(stat= "identity", position = "dodge") +
    facet_grid(organ ~ large_summarized_class,
               scale = "free_x", space = "free_x",
               labeller = label_wrap_gen(5)) +
    theme_bw() +
    theme(axis.text.x =
              element_text(angle = 90, hjust = 1, vjust = 0.5),
          legend.position = "bottom")
```

```{r}

```
 
 
 %>%
    dplyr::select(ID, organ, cluster, Gene_name, share, Homology_type) %>%
    distinct %>%
    mutate(cell_label = paste(organ, cluster, sep = "_")) %>%
    group_by(cell_label, ID) %>%
    dplyr::count(share) %>%
    ungroup() %>% 
    tidyr::complete(cell_label, share) %>%
    mutate(organ = str_split(cell_label, pattern = "_", simplify = TRUE)[,1]) -> count_marker_gene_per_share

count_marker_gene_per_share %>%
    inner_join(hs_mm_marker_disease.do, by = c("ID" = "DO_Disease_ID"))  %>%
    ungroup() %>%
    tidyr::complete(cell_label, organ, ID, share) %>%
    mutate(n = replace_na(n, 0)) %>%
    ggplot(aes(x = ID, y = n, fill = share)) +
    geom_bar(stat ="identity", position = "dodge") +
    facet_grid(cell_label ~ .) +
    theme_bw() +
    theme(axis.text.x =
              element_text(angle = 90, hjust = 1, vjust = 0.5),
          legend.position = "bottom")
