---
title: "Untitled"
output: html_document
date: "2022-11-28"
---

```{r}
library(tidyverse)
library(scAlign)
library(Seurat)
library(SingleCellExperiment)
library(gridExtra)
library(cowplot)
```

ヒト種特異的な遺伝子の再選択
これまでの遺伝子対応表には同じ遺伝子名で異なる対応関係がアノテーションされている場合があったため、そういった遺伝子は種特異的な遺伝子から外したい
```{r}
read_tsv("Gene_name/Hs.Gene_name.homology_type.txt") -> Hs.gene.homology
# readRDS(file="/Users/kariyayama/Labolatory/JST_Mouse_Human/Tabula/Tabula_sapiens/TS_liver.rds") -> Hs_liver

# Tabula sapiensで使われている遺伝子名
Hs_liver@assays$RNA@counts %>%
  rownames() -> TS_genes

# Tabula sapiensで使われている遺伝子名と遺伝子対応関係表を結合
data.frame(Gene_name = TS_genes) %>%
  inner_join(Hs.gene.homology,
             by = "Gene_name") %>%
  arrange(Gene_name) -> Hs.gene.homology.TS

# 保存
write.table(Hs.gene.homology.TS,
            "Gene_name/Hs.Gene_name.homology_type.Tabula_sapiens.txt",
            quote = F,
            sep = "\t",
            row.names = F)
```

Tabula sapiensで使われている遺伝子の対応関係表からどの遺伝子IDでも対応遺伝子が見つからない遺伝子を抽出
```{r}
Hs.gene.homology.TS %>%
  dplyr::filter(is.na(Mouse_homology_type)) -> Hs.gene.homology.TS.ss
Hs.gene.homology.TS %>%
  dplyr::filter(!is.na(Mouse_homology_type)) -> Hs.gene.homology.TS.homolog
Hs.gene.homology.TS.ss %>%
  inner_join(Hs.gene.homology.TS.homolog, by = "Gene_name") %>%
  dplyr::select(Gene_name)  -> Hs.gene.homology.TS.not_ss
Hs.gene.homology.TS.ss %>%
  anti_join(Hs.gene.homology.TS.not_ss,
            by = "Gene_name") -> Hs.gene.homology.TS.ss
write.table(Hs.gene.homology.TS.ss,
            "Gene_name/Hs.Gene_name.species_specific.v2.txt",
            quote = F,
            sep = "\t",
            row.names = F)
```

マウス種特異的な遺伝子の再選択
これまでの遺伝子対応表には同じ遺伝子名で異なる対応関係がアノテーションされている場合があったため、そういった遺伝子は種特異的な遺伝子から外したい
```{r}
read_tsv("Gene_name/Mm.Gene_name.homology_type.txt") -> Mm.gene.homology
# readRDS(file="/Users/kariyayama/Labolatory/JST_Mouse_Human/Tabula/Tabula_muris/tabula-muris/00_data_ingest/00_facs_raw_data/FACS/Liver.RDS") -> Mm_liver
Mm_liver@assays$RNA@counts %>%
  rownames() -> TM_genes
data.frame(Gene_name = TM_genes) %>%
  inner_join(Mm.gene.homology,
             by = "Gene_name") %>%
  arrange(Gene_name) -> Mm.gene.homology.TM

write.table(Mm.gene.homology.TM,
            "Gene_name/Mm.Gene_name.homology_type.Tabula_muris.txt",
            quote = F,
            sep = "\t",
            row.names = F)
```

Tabula murisで使われている遺伝子の対応関係表からどの遺伝子IDでも対応遺伝子が見つからない遺伝子を抽出
```{r}
Mm.gene.homology.TM %>%
  dplyr::filter(is.na(Human_homology_type)) -> Mm.gene.homology.TM.ss
Mm.gene.homology.TM %>%
  dplyr::filter(!is.na(Human_homology_type)) -> Mm.gene.homology.TM.homolog
Mm.gene.homology.TM.ss %>%
  inner_join(Mm.gene.homology.TM.homolog, by = "Gene_name") %>%
  dplyr::select(Gene_name)  -> Mm.gene.homology.TM.not_ss
Mm.gene.homology.TM.ss %>%
  anti_join(Mm.gene.homology.TM.not_ss,
            by = "Gene_name") -> Mm.gene.homology.TM.ss
write.table(Mm.gene.homology.TM.ss,
            "Gene_name/Mm.Gene_name.species_specific.v2.txt",
            quote = F,
            sep = "\t",
            row.names = F)
```

Human_mouse_liverで行った種特異的な遺伝子の合計発現量の可視化
統合データ・ラベルされたオブジェクトの読み込み
```{r}
# Mm_liver.picked <- readRDS("Liver/Mm_liver.picked.rds")
# Hs_liver.picked <- readRDS("Liver/Hs_liver.picked.rds")
# Hs_Mm.seurat.combined <- readRDS(Hs_Mm.seurat.combined,
#                                 file = "Liver/Hs_Mm.seurat.combined.rds")
DimPlot(Hs_Mm.seurat.combined,
        split.by = "sp.ident",
        group.by = "cell_ontology_class",
        repel = T, label = T)  + NoLegend() -> p1
# DimPlot(Hs_Mm.seurat.combined,
#         split.by = "sp.ident",
#         group.by = "ident",
#         repel = T, label = T)  + NoLegend() -> p1.ident
```
ヒト・マウスの種特異的な遺伝子のリストを用意
```{r}
# 系統特異的な遺伝子リストを用意
Hs_only <- read_tsv("Gene_name/Hs.Gene_name.species_specific.v2.txt",
                    show_col_types = FALSE)
Mm_only <- read_tsv("Gene_name/Mm.Gene_name.species_specific.v2.txt",
                    show_col_types = FALSE)
```

ヒトとマウスについて、マーカー遺伝子のうち種特異的なものを抽出
```{r}
# マウス特異的マーカー遺伝子
Mm_liver.picked.markers %>%
  inner_join(Mm_only, by = c("gene" = "Gene_name")) %>%
  filter(p_val_adj < 0.05) %>%
  write.table(file = "Liver/Hs_Mm_liver.seurat_integrate.celltype_pick.Mm_only.markergene.tsv",
              sep = "\t",
              row.names = F,
              quote = F)

# ヒトオーソログをもつマウスマーカー遺伝子
Mm_liver.picked.markers %>%
  anti_join(Mm_only, by = c("gene" = "Gene_name")) %>%
  filter(p_val_adj < 0.05) %>%
  write.table(file = "Liver/Hs_Mm_liver.seurat_integrate.celltype_pick.Mm_ortholog.markergene.tsv",
              sep = "\t",
              row.names = F,
              quote = F)

# ヒト特異的マーカー遺伝子
Hs_liver.picked.markers %>%
  inner_join(Hs_only, by = c("gene" = "Gene_name")) %>%
  filter(p_val_adj < 0.05) -> Hs_liver.picked.markers.SS_gene

Hs_liver.picked.markers.SS_gene %>%
  write.table(file = "Liver/Hs_Mm_liver.seurat_integrate.celltype_pick.Hs_only.markergene.tsv",
              sep = "\t",
              row.names = F,
              quote = F)

# マウスオーソログを持つヒトマーカー遺伝子
Hs_liver.picked.markers %>%
  anti_join(Hs_only, by = c("gene" = "Gene_name")) %>%
  filter(p_val_adj < 0.05) -> Hs_liver.picked.markers.orth_gene

Hs_liver.picked.markers.orth_gene %>%
  write.table(file = "Liver/Hs_Mm_liver.seurat_integrate.celltype_pick.Hs_ortholog.markergene.tsv",
              sep = "\t",
              row.names = F,
              quote = F)
```

種特異的な遺伝子の発現量をプロット
```{r}
Hs_Mm.seurat.combined@meta.data %>%
  rownames_to_column(var = "Cell") ->
  Hs_Mm.seurat.meta.data

# 系統特異的な遺伝子の合計を細胞ごとに計算
Mm_liver@assays$RNA@data %>%
  as.data.frame %>%
  rownames_to_column("Gene") %>%
  inner_join(Mm_only, by = c("Gene" = "Gene_name")) %>%
  column_to_rownames(var = "Gene") %>%
  apply(2, sum) -> Mm_liver.Mm_only.sum

Hs_liver@assays$RNA@data %>%
  as.data.frame %>%
  rownames_to_column("Gene") %>%
  inner_join(Hs_only, by = c("Gene" = "Gene_name")) %>%
  column_to_rownames(var = "Gene") %>%
  apply(2, sum) -> Hs_liver.Hs_only.sum

liver_only <- c(Mm_liver.Mm_only.sum, Hs_liver.Hs_only.sum)
data.frame(Cell = names(liver_only),
           SUM = liver_only) %>%
  as_tibble %>%
  inner_join(Hs_Mm.seurat.meta.data, by = "Cell") -> liver.only.sum.meta

# 系統特異的な遺伝子の合計をプロット
liver.only.sum.meta %>%
    ggplot(aes(x = seurat_clusters, y = SUM, fill = seurat_clusters)) +
    geom_violin() +
    facet_wrap( ~ sp.ident) +
    geom_dotplot(binaxis = "y", stackdir = "center", binwidth = 5) +
    theme_classic() +
    NoLegend() -> p5.sum
p1 + p1.ident + p5.sum -> p6
```
マウスの場合、種特異的な遺伝子の発現量が大きいHepatocyte(Ident: 3)でマーカー遺伝子と判断される種特異的な遺伝子の数が多い

マーカー遺伝子のうち、種特異的な遺伝子が占める割合をプロットする
```{r}
# マウス
Mm_liver.picked.markers %>%
  inner_join(Mm_only, by = c("gene" = "Gene_name")) %>%
  filter(p_val_adj < 0.05) %>%
  dplyr::count(cluster) -> Mm_liver.picked.only_markers.counts
Mm_liver.picked.markers %>%
  filter(p_val_adj < 0.05) %>%
  dplyr::count(cluster) %>%
  inner_join(Mm_liver.picked.only_markers.counts,
             by = "cluster") %>%
  mutate(sp = "Mm") %>%
  mutate(ratio = n.y / n.x) -> Mm_liver.picked.counts

# ヒト
Hs_liver.picked.markers %>%
  inner_join(Hs_only, by = c("gene" = "Gene_name")) %>%
  filter(p_val_adj < 0.05) %>%
  dplyr::count(cluster) -> Hs_liver.picked.only_markers.counts
Hs_liver.picked.markers %>%
  filter(p_val_adj < 0.05) %>%
  dplyr::count(cluster) %>%
  inner_join(Hs_liver.picked.only_markers.counts,
             by = "cluster")  %>%
  mutate(sp = "Hs") %>%
  mutate(ratio = n.y / n.x) -> Hs_liver.picked.counts

# ヒトとマウスの統合
Hs_liver.picked.counts %>%
  bind_rows(Mm_liver.picked.counts) %>%
  ggplot(aes(x=cluster, y = n.y, fill = cluster)) +
    geom_bar(stat = "identity") +
    facet_wrap( ~ sp) +
    theme_classic() +
    NoLegend() -> p7
p6 + p7 -> p8
p8 %>%
  ggsave("Liver/Hs_Mm_liver.seurat_integrate.celltype_pick.only_gene.png", 
         ., width = 12, height = 20)

```

Ratioで計算する
```{r}
# 系統特異的な遺伝子の合計を細胞ごとに計算
Mm_liver@assays$RNA@data %>%
  as.data.frame %>%
  apply(2, sum) -> Mm_liver.all.sum

Hs_liver@assays$RNA@data %>%
  as.data.frame %>%
  apply(2, sum) -> Hs_liver.all.sum

liver_all <- c(Mm_liver.all.sum, Hs_liver.all.sum)
data.frame(Cell = names(liver_all),
           SUM.all = liver_all) %>%
  as_tibble %>%
  inner_join(liver.only.sum.meta, by = "Cell") -> liver.sum.meta

# 系統特異的な遺伝子の合計をプロット
liver.sum.meta %>%
    mutate(ratio = SUM / SUM.all) %>%
    ggplot(aes(x = seurat_clusters, y = ratio, fill = seurat_clusters)) +
    geom_violin() +
    facet_wrap( ~ sp.ident) +
    geom_dotplot(binaxis = "y", stackdir = "center", binwidth = 0.0005) +
    theme_classic() +
    NoLegend() -> p5.ratio

# ヒトとマウスの割合をプロット
Hs_liver.picked.counts %>%
  bind_rows(Mm_liver.picked.counts) %>%
  ggplot(aes(x=cluster, y = ratio, fill = cluster)) +
    geom_bar(stat = "identity") +
    facet_wrap( ~ sp) +
    theme_classic() +
    NoLegend() -> p7.ratio
p1 + p1.ident + p5.ratio + p7.ratio-> p8.ratio
p8.ratio %>%
  ggsave("Liver/Hs_Mm_liver.seurat_integrate.celltype_pick.only_gene.ratio.png", 
         ., width = 12, height = 20)
```

