---
title: "specificity_index"
output: html_document
date: "2022-11-18"
---
肝臓細胞クラスターで特異的に発現しているヒト特異的遺伝子とマウス特異的遺伝子のGO解析
```{r}
library(rbioapi)
library(tidyverse)
library(scAlign)
library(Seurat)
library(SingleCellExperiment)
library(gridExtra)
library(cowplot)
setwd("/Users/kariyayama/Labolatory/JST_Mouse_Human/Tabula/Integration")
```

ヒトのマーカー遺伝子
BP: GO:0008150
MF: GO:0003674
CC: GO:0005575
```{r}
# マーカー検出結果の取得
read_tsv("Liver/Version0/Hs_Mm_liver.seurat_integrate.celltype_pick.Hs_markergene.version0.tsv",
         show_col_types = FALSE) -> Hs_all.marker
read_tsv("Liver/Version0/Hs_Mm_liver.seurat_integrate.celltype_pick.Hs_only.markergene.version0.tsv",
         show_col_types = FALSE) -> Hs_only.marker
read_tsv("Liver/Version0/Hs_Mm_liver.seurat_integrate.celltype_pick.Hs_ortholog.markergene.version0.tsv",
         show_col_types = FALSE) -> Hs_ortholog.marker
```

定数定義
```{r}
test_type <- "FISHER"
cutoff <- 0.05
hsapiens <- 9606
mmuculus <- 10090
BP <- "GO:0008150"
```

解析用の関数
```{r}
filter_cluster <- function(x, ident){
  x  %>%
    filter(cluster == ident) %>%
    select(gene) %>% unlist %>% as.character()
}

createGOanalysis_outfile <- function(sp, ident, testtype = "FISHER", target)
  paste("Liver/PANTHER/", sp, "/", sp, "_PANTHER_ident_",
        cluster.id, "_", testtype, "__BP_", target, "_gene.version0.tsv", sep="")
```

```{r}
# Ident 0
Hs_all.marker %>%
  filter_cluster(., 3) -> Hs_all.ident3
Hs_only.marker %>%
  filter_cluster(., 0) -> Hs_only.ident0

rba_panther_enrich(Hs_only.ident0, hsapiens, BP,
                   cutoff = cutoff) -> Hs.ident0.all_result

rba_panther_enrich(Hs_only.ident0, hsapiens, BP,
                   ref_genes = Hs_all.ident0, ref_organism = hsapiens,
                   cutoff = cutoff) -> Hs.ident0_result
```

全てのIdentに対して実行
```{r}
# ヒトの細胞クラスターの一覧を取得する
Hs_cluster_list <- Hs_only.marker %>%
  select(cluster) %>% unlist %>% 
  as.character %>% unique 

for(cluster.id in Hs_cluster_list){
  print(cluster.id)
  
  # Read data
  Hs_all.marker %>%
    filter_cluster(., cluster.id) -> Hs_all.ident
  Hs_only.marker %>%
    filter_cluster(., cluster.id) -> Hs_only.ident
  Hs_ortholog.marker %>%
    filter_cluster(., cluster.id) -> Hs_ortholog.ident

  # ヒト特異的遺伝子のGO解析
  # 比較対象が全遺伝子
  rba_panther_enrich(Hs_only.ident, hsapiens, BP, test_type = test_type,
                     cutoff = cutoff) -> Hs.ident.all_result
  outfile_all <- createGOanalysis_outfile("Hs", cluster.id,
                                          test_type, "all")
  write.table(Hs.ident.all_result$result, file = outfile_all,
              quote = F, row.names = F, sep="\t")
  # 比較対象がマーカー遺伝子
  rba_panther_enrich(Hs_only.ident, hsapiens, BP, test_type = test_type,
                     ref_genes = Hs_all.ident, ref_organism = hsapiens,
                     cutoff = cutoff) -> Hs.ident_result
  outfile_marker <- createGOanalysis_outfile("Hs", cluster.id,
                                             test_type, "marker")
  write.table(Hs.ident_result$result, file = outfile_marker,
              quote = F, row.names = F, sep="\t")

    # マウスオーソログを持つヒト遺伝子のGO解析
  rba_panther_enrich(Hs_ortholog.ident, hsapiens, BP, test_type = test_type,
                     cutoff = cutoff) -> Hs_ortholog.ident
  outfile_ortholog <- createGOanalysis_outfile("Hs", cluster.id,
                                             test_type, "ortholog")
  write.table(Hs_ortholog.ident$result, file = outfile_ortholog,
              quote = F, row.names = F, sep="\t")
}
```

マウスのマーカー遺伝子
```{r}
# マーカー検出結果の取得
read_tsv("Liver/Hs_Mm_liver.seurat_integrate.celltype_pick.Mm_markergene.tsv",
         show_col_types = FALSE) -> Mm_all.marker
read_tsv("Liver/Hs_Mm_liver.seurat_integrate.celltype_pick.Mm_only.markergene.tsv",
         show_col_types = FALSE) -> Mm_only.marker
# ヒトオーソログを持つマウス遺伝子のGO解析
read_tsv("Liver/Hs_Mm_liver.seurat_integrate.celltype_pick.Mm_ortholog.markergene.tsv",
         show_col_types = FALSE) -> Mm_ortholog.marker
```
全てのIdentに対して実行
```{r}
# クラスターの名前一覧を取得する
Mm_cluster_list <- Mm_only.marker %>%
  select(cluster) %>%
  unlist %>% 
  as.character %>%
  unique

for(cluster.id in Mm_cluster_list){
  print(cluster.id)
  
  # Read data
  Mm_all.marker %>%
    filter_cluster(., cluster.id) -> Mm_all.ident
  Mm_only.marker %>%
    filter_cluster(., cluster.id) -> Mm_only.ident
  Mm_ortholog.marker %>%
    filter_cluster(., cluster.id) -> Mm_ortholog.ident
  
  # outfile
  outfile_all      <- createGOanalysis_outfile("Mm", cluster.id,
                                               test_type, "all")
  outfile_marker   <- createGOanalysis_outfile("Mm", cluster.id,
                                               test_type, "marker")
  outfile_ortholog <- createGOanalysis_outfile("Mm", cluster.id,
                                               test_type, "ortholog")
  
  # GO解析
  rba_panther_enrich(Mm_all.ident, mmuculus, BP, test_type = test_type,
                     cutoff = cutoff) -> Mm.ident.all_result
  write.table(Mm.ident.all_result$result, file = outfile_all, quote = F,
              row.names = F, sep="\t")
  
  rba_panther_enrich(Mm_only.ident, mmuculus, BP, test_type = test_type,
                     ref_genes = Mm_all.ident, ref_organism = mmuculus,
                     cutoff = cutoff) -> Mm.ident_result
  write.table(Mm.ident_result$result, file = outfile_marker, quote = F,
              row.names = F, sep="\t")
  
  # ヒトオーソログを持つマウス遺伝子のGO解析
  rba_panther_enrich(Mm_ortholog.ident, mmuculus, BP, test_type = test_type,
                     cutoff = cutoff) -> Mm_ortholog.ident
  write.table(Mm_ortholog.ident$result,
              file = outfile_ortholog, quote = F, row.names = F, sep="\t")
}
```

GO解析の結果をプロットして可視化
1細胞クラスターについてやってみる
```{r}
make_GO_short <- function(acc, short){
    result <- c()
    for(i in 1:length(acc)){
          if(nchar(unlist(short[i])) > 20){
              result <- c(result,
                          paste(substr(short[i], 1, 17),
                                "... (", acc[i], ")",
                                sep=""))
          }else{
              result <- c(result,
                          paste(short[i], " (", acc[i], ")",
                                sep=""))
          }
    }
    return(result)
}
```

```{r}
# Ident 0
read_tsv("Liver/PANTHER/Hs/Hs_PANTHER_ident_3_FISHER__BP_all_gene.tsv",
         show_col_types = FALSE) -> Hs_all.ident3
read_tsv("Liver/PANTHER/Mm/Mm_PANTHER_ident_3_FISHER__BP_all_gene.tsv",
         show_col_types = FALSE) -> Mm_all.ident3

Hs_all.ident3 %>%
  right_join(Mm_all.ident3, by = c("term.id", "term.label")) %>%
  select(c(term.id, term.label, fdr.x, fdr.y)) %>%
  dplyr::rename(human = fdr.x, mouse = fdr.y) %>%
  pivot_longer(cols=-c(term.id, term.label), 
               names_to = "SP", values_to = "FDR") %>%
  mutate(fdr_log = -log10(replace(FDR, is.na(FDR), 1))) %>%
  mutate(GO_short_acc = make_GO_short(term.id, term.label)) %>%
  dplyr::select(c(GO_short_acc, fdr_log, SP)) -> out_table3

out_table3 %>%
  ggplot(aes(x=GO_short_acc, y = fdr_log, fill = SP)) +
    geom_bar(stat = "identity", position = "dodge") +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 90,
                                     vjust = 0.5,
                                     hjust = 0)
          )
```

GO解析の結果を読み込み
```{r}
read_data <- function(sp, cluster, type = "all"){
  read_tsv(paste("Liver/PANTHER/", sp, "/", sp, "_PANTHER_ident_",
                 cluster, "_FISHER__BP_", type, "_gene.version0.tsv",
                 sep = ""),
           show_col_types = FALSE) %>%
    filter(plus_minus == "+")
}

# データの加工、NAを1に変換する
na_to_1 <- function(x) as.numeric(replace(x, is.na(x), 1))

# plot
plot_gocompare <- function(outtable)
    out_table %>%
      mutate(GO_short_acc = make_GO_short(term.id, term.label)) %>%
      ggplot() +
        theme_classic() +
        geom_hline(aes(yintercept = threshold), linetype = "dashed") +
        theme(axis.text.x = plot_para)

# 定数
threshold <- -log10(0.05)
plot_para <- element_text(angle = 90, vjust = 0.5, hjust = 0)
height = 16
width <- function(x)  x * 0.05 + 5
```  

```{r}
merge_different_species <- function(hs, mm)
    hs %>%
      full_join(mm, by = c("term.id", "term.label")) %>%
      dplyr::rename(human = fdr.x, mouse = fdr.y) %>%
      mutate(human = na_to_1(human), mouse = na_to_1(mouse)) %>%
      dplyr::select(c("term.id", "term.label", "human", "mouse")) %>%
      pivot_longer(cols=-c(term.id, term.label), 
                   names_to = "SP", values_to = "FDR") %>%
      mutate(fdr_log = -log10(FDR))
```

全ヒト ・マウス：マウス共通の細胞クラスターについて、種特異的なマーカー遺伝子vsそれ以外のGO解析の結果を可視化する
```{r}
for(ident in Mm_cluster_list){
  read_data("Hs", ident) -> Hs_all.ident
  read_data("Mm", ident) -> Mm_all.ident
  outfile <- paste("Liver/PANTHER/Comparison/All/compare_ident_",
               ident, "_human-mouse.version0.png", sep= "")
  
  merge_different_species(Hs_all.ident,
                          Mm_all.ident) -> out_table
  
  if(dim(out_table)[1] > 0){
    plot_gocompare(out_table) +
      geom_bar(aes(x=GO_short_acc, y = fdr_log, fill = SP),
               stat = "identity", position = "dodge") -> p1
    
    ggsave(outfile, p1, width = width(dim(out_table)[1]),
           height = height, limitsize = FALSE)
  }
}
```

全ヒト・マウス：マウス共通の細胞クラスターについて、種特異的なマーカー遺伝子vsそれ以外のマーカー遺伝子のGO解析の結果を可視化する
```{r}
for(ident in Mm_cluster_list){
  read_data("Hs", ident, "marker") -> Hs_all.ident
  read_data("Mm", ident, "marker") -> Mm_all.ident
  outfile <- paste("Liver/PANTHER/Comparison/Marker_gene/compare_ident_",
                 ident, "_human-mouse.version0.png", sep= "")
    
  merge_different_species(Hs_all.ident,
                          Mm_all.ident) -> out_table

  if(dim(out_table)[1] > 0){
    plot_gocompare(out_table) +
      geom_bar(aes(x=GO_short_acc, y = fdr_log, fill = SP),
               stat = "identity", position = "dodge") -> p1
    ggsave(outfile, p1, width = width(dim(out_table)[1]), height = height)
  }
}
```

```{r}
# homology_typeごと
merge_different_homology_type <- function(all, ortholog)
    all %>%
        full_join(ortholog, by = c("term.id", "term.label")) %>%
        dplyr::rename(SS = fdr.x, ortholog = fdr.y) %>%
        mutate(SS = na_to_1(SS), ortholog = na_to_1(ortholog)) %>%
        dplyr::select(c("term.id", "term.label", "SS", "ortholog")) %>%
        pivot_longer(cols=-c(term.id, term.label), 
                     names_to = "ortholog_type", values_to = "FDR") %>%
        mutate(fdr_log = -log10(FDR))
```

ヒトのオーソログがある遺伝子と種特異的遺伝子間の比較
```{r}
for(ident in 0:14){
  read_data("Hs", ident, "all") -> Hs_all.ident
  read_data("Hs", ident, "ortholog") -> Hs_ortholog.ident
  outfile <- paste("Liver/PANTHER/Comparison/All/Hs_compare_ident_",
                   ident, "_SS-ortholog.version0.png", sep= "")
  merge_different_homology_type(Hs_all.ident,
                                Hs_ortholog.ident) -> out_table
  
  if(dim(out_table)[1] > 0){
    plot_gocompare(out_table) +
      geom_bar(aes(x=GO_short_acc, y = fdr_log, fill = ortholog_type),
               stat = "identity", position = "dodge") -> p1

  ggsave(outfile, p1, width = width(dim(out_table)[1]), height = height,
         limitsize = FALSE)
  }
}
```

マウスのオーソログがある遺伝子と種特異的遺伝子間の比較
```{r}
for(ident in Mm_cluster_list){
  read_data("Mm", ident, "all") -> Mm_all.ident
  read_data("Mm", ident, "ortholog") -> Mm_ortholog.ident
  outfile <- paste("Liver/PANTHER/Comparison/All/Mm_compare_ident_",
                 ident, "_SS-ortholog.version0.png", sep= "")

  merge_different_homology_type(Mm_all.ident,
                                Mm_ortholog.ident) -> out_table

  if(dim(out_table)[1] > 0){  
    plot_gocompare(out_table) +
      geom_bar(aes(x=GO_short_acc, y = fdr_log, fill = ortholog_type),
               stat = "identity", position = "dodge") -> p1
    ggsave(outfile, p1, width = width(dim(out_table)[1]), height = height,
           limitsize = FALSE)
  }
}
```
