---
title: "Count_marker_gene"
output: html_document
date: "2023-01-09"
---
```{r}
curdir <- "~/Labolatory/JST_Mouse_Human/Tabula/Integration"
# renv::restore()
setwd(curdir)
library(here)
set_here(curdir)

source(paste(curdir, "integration_analysis_function.R", sep = "/"))
organ <- "Liver"
```

ENSEMBL synonym
```{r}
read_tsv("Gene_name/20230112_Mm_Hs_gene_name_release104.txt",
         show_col_types = FALSE) -> Mm_ortholog_raw
# Reference_genome
read_tsv("Gene_name/Mm_MM10-PLUS_genes.list") -> MM10_PLUS # 23,433 genes

# ENSEMBL ID - Refseq
ensembl_refseq <- read_tsv("Gene_name/ENSEMBL/20230120_Mm_chromosome_release104.txt",
                           show_col_types = FALSE)
mgi_ensembl <- read_tsv("Gene_name/MGI/MRK_ENSEMBL.rpt",
                        show_col_types = FALSE)

# Refseq IDを使って
inner_join(ensembl_refseq,
           MM10_PLUS,
           by = c("RefSeq_mRNA_ID" = "Refseq")) %>%
  select(c("Gene_stable_ID", "Gene_name.y")) %>%
  distinct() -> refseq_mm10_mrna# 20,200 gene names, 20216 gene ID
inner_join(ensembl_refseq,
           MM10_PLUS,
           by = c("RefSeq_ncRNA_ID" = "Refseq")) %>%
  select(c("Gene_stable_ID", "Gene_name.y")) %>%
  distinct() -> refseq_mm10_lnrna # 2,712 genes, 2713 gene ID

bind_rows(refseq_mm10_mrna, refseq_mm10_lnrna) %>%
  inner_join(Mm_seurat_gene, .,
             by = c("Mouse_gene_name" = "Gene_name.y")) %>%
  inner_join(Mm_ortholog_raw, by = "Gene_stable_ID") %>%
  select(c("Gene_stable_ID", "Mouse_gene_name",
           "Human_gene_stable_ID", "Human_homology_type")) %>%
  distinct %>%
  mutate(Human_homology_type = 
           replace(Human_homology_type,
                   is.na(Human_homology_type),
                   "SS")) -> Mm_gene_homology_type
Mm_gene_homology_type %>%
  write.table("Gene_name/20230120_Mm_Hs_refseq_muris.tsv",
              sep = "\t", quote = F, row.names = F)
Mm_gene_homology_type %>%
  select(c(Gene_stable_ID, Mouse_gene_name)) %>%
  distinct %>%
  dplyr::count(Mouse_gene_name) %>%
  filter(n > 1) -> Mm_not_uniq
Mm_gene_homology_type %>%
  anti_join(Mm_not_uniq, by = "Mouse_gene_name") -> Mm_gene_homology_type.uniq
```

Human genes
```{r}
# ENSEMBL homology table
read_tsv("Gene_name/20230112_Hs_Mm_gene_name_release104.txt",
         show_col_types = FALSE) %>%
  select(c("Gene_stable_ID", "Gene_name",
           "Mouse_gene_stable_ID", "Mouse_homology_type")) %>%
  mutate(Mouse_homology_type = replace(Mouse_homology_type,
                                       is.na(Mouse_homology_type),
                                       "SS")) %>%
  distinct -> Hs_ortholog_raw
# gencodeの表の読み込み
read_tsv("Gene_name/gencode.v30.transcripts.tsv",
         show_col_types = FALSE) -> Hs_gencode

# Tabula sapiensの遺伝子のうち、gencodeに含まれていないもの
read_tsv("Gene_name/gencode.v30.transcripts.tsv",
         show_col_types = FALSE) %>%
  full_join(Hs_seurat_gene, by = "Gene_name") %>%
  filter(is.na(ENSEMBLE_gene_ID)) # 1,554遺伝子はTabula sapiensにあるが、gencodeにない
```
変数に変換する
```{r}
# gencodeとTabula sapiens共通の遺伝子をENSEMBL Biomartと結合させた
Hs_ortholog_raw %>%
  inner_join(Hs_gencode,
            by = c("Gene_stable_ID" = "ENSEMBL_gene_ID")) %>%
  select(c("Gene_stable_ID", "Gene_name.x", "Gene_name.y",
           "Mouse_gene_stable_ID", "Mouse_homology_type")) %>%
  distinct -> Hs_g_t
Hs_g_t %>%
  inner_join(Hs_seurat_gene, .,
            by = c("Gene_name" = "Gene_name.x")) %>%
  select(c("Gene_stable_ID", "Gene_name",
           "Mouse_gene_stable_ID", "Mouse_homology_type")) -> Hs_g_t.x
Hs_g_t %>%
  inner_join(Hs_seurat_gene, .,
            by = c("Gene_name" = "Gene_name.y")) %>%
  select(c("Gene_stable_ID", "Gene_name",
           "Mouse_gene_stable_ID", "Mouse_homology_type"))-> Hs_g_t.y # 58,604遺伝子は共通
bind_rows(Hs_g_t.x, Hs_g_t.y) %>%
  distinct -> Hs_gene_homology_type
Hs_gene_homology_type %>%
  write.table("Gene_name/20230119_Hs_Mm_genecode_sapiens.tsv",
              sep = "\t", quote = F, row.names = F)
Hs_gene_homology_type %>%
  select(c(Gene_stable_ID, Gene_name)) %>%
  distinct %>%
  dplyr::count(Gene_name) %>%
  filter(n > 1) -> Hs_not_uniq
Hs_gene_homology_type %>%
  anti_join(Hs_not_uniq, by = "Gene_name") -> Hs_gene_homology_type.uniq
```

ヒト-マウスを統合
```{r}
full_join(Hs_gene_homology_type.uniq,
          Mm_gene_homology_type.uniq,
          by = c("Gene_stable_ID" = "Human_gene_stable_ID",
                 "Mouse_gene_stable_ID" = "Gene_stable_ID")) %>%
  write.table("Gene_name/20230120_Hs-Mm_integrated_full.tsv",
              sep = "\t", quote = F, row.names = F)
```
homology_type情報を追加
```{r}
# one2one ortholog, but ambigous
homology %>%
  filter(Mouse_homology_type == "ortholog_one2one" &
           Mouse_homology_type == Human_homology_type) %>%
    dplyr::count(Gene_name) %>%
    filter(n > 1) -> Hs_Mm.notuniq

# strinct one2one ortholog
homology %>%
  filter(Mouse_homology_type == "ortholog_one2one" &
           Mouse_homology_type == Human_homology_type) %>%
  anti_join(Hs_Mm.notuniq, by = "Gene_name") %>%
  mutate(Homology_type = "ortholog_one2one") -> Hs_Mm.one2one

# one2many ortholog
homology %>%
  filter(Mouse_homology_type == "ortholog_one2many" &
           Mouse_homology_type == Human_homology_type) %>%
  anti_join(Hs_Mm.notuniq, by = "Gene_name") %>%
  mutate(Homology_type = "ortholog_one2many") -> Hs_Mm.one2many

# many2many ortholog
homology %>%
  filter(Mouse_homology_type == "ortholog_many2many" &
           Mouse_homology_type == Human_homology_type) %>%
  anti_join(Hs_Mm.notuniq, by = "Gene_name") %>%
  mutate(Homology_type = "ortholog_many2many") -> Hs_Mm.many2many

# Human specific
homology %>%
    filter(Mouse_homology_type == "SS" & !is.na(Gene_name)) %>%
    distinct() %>%
    mutate(Homology_type = "Hs_SS") -> Hs_only

# Mouse specific
homology %>%
    filter(Human_homology_type == "SS" & !is.na(Mouse_gene_name)) %>%
    distinct() %>%
    mutate(Homology_type = "Mm_SS")  -> Mm_only

# Ambigous
homology %>%
    mutate(Human_homology_type = 
                        replace(Human_homology_type,
                                is.na(Human_homology_type),
                                "SS")) %>%
    mutate(Mouse_homology_type = 
               replace(Mouse_homology_type,
                       is.na(Mouse_homology_type),
                       "SS")) %>%
    filter(Mouse_homology_type != Human_homology_type) -> Hs_Mm.ambigous
homology %>%
  inner_join(Hs_Mm.notuniq, by = "Gene_name") %>% 
  select(c(1:6)) %>%
  bind_rows(Hs_Mm.ambigous) %>%
  mutate(Homology_type = "Ambigous") -> Hs_Mm.ambigous
# Retired
homology %>%
  filter(is.na(Mouse_homology_type) & is.na(Mouse_gene_stable_ID)) %>%
  mutate(Homology_type = "Retired") -> Hs.retired
homology %>%
    filter(is.na(Human_homology_type) & is.na(Gene_stable_ID)) %>%
    mutate(Homology_type = "Retired") -> Mm.retired # No_gene

bind_rows(Hs_Mm.one2one, Hs_Mm.one2many, Hs_Mm.many2many,
          Hs_only, Mm_only, Hs_Mm.ambigous, Hs.retired) %>%
  write.table("Gene_name/20230120_Hs-Mm_integrated_full.version2.tsv",
              sep = "\t", quote = F, row.names = F)
  
``` 
MM10 Plusを入れて確認した
-> 直接Gene nameをENSEMBL stable IDに変換することはできなかった。ただしgtfファイルはあるため、無理やり復元させることはできそう


```{r}

human_mouse_ortholog_raw %>%
    inner_join(human_orthology,
               by = c("Gene_name",
                      "Homology_type" = "Mouse_homology_type.x")) %>%
    left_join(mouse_orthology,
              by = c("Mouse_gene_name" = "Gene_name",
                     "Homology_type" = "Human_homology_type.x")) %>%
    arrange(Gene_name) %>%
    dplyr::rename(Human_gene_name = Gene_name) %>%
  select(c(1:3)) -> human_mouse_ortholog_human

mouse_human_ortholog_raw %>%
    inner_join(mouse_orthology,
               by = c("Gene_name",
                      "Homology_type" = "Human_homology_type.x")) %>%
    left_join(human_orthology,
              by = c("Human_gene_name" = "Gene_name",
                     "Homology_type" = "Mouse_homology_type.x")) %>%
    arrange(Gene_name) %>%
    dplyr::rename(Mouse_gene_name = Gene_name) %>%
    select(c(1:3)) -> human_mouse_ortholog_mouse

bind_rows(human_mouse_ortholog_human, human_mouse_ortholog_mouse) %>%
  distinct %>%
  write.table("Gene_name/20230112_human_mouse_orthology.txt",
              sep = "\t", quote = F, row.names = F)

human_mouse_ortholog_raw %>%
    inner_join(human_orthology,
               by = c("Gene_name", "Mouse_homology_type" = "Mouse_homology_type.x")) %>%
    left_join(mouse_orthology,
              by = c("Mouse_gene_name" = "Gene_name",
                     "Mouse_homology_type" = "Human_homology_type.x")) %>%
  inner_join(Mm_seurat_gene, by = "Mouse_gene_name") %>%
  inner_join(Hs_seurat_gene, by = "Gene_name") %>%
  select(c(1:3)) -> human_mouse_ortholog
```

マーカー遺伝子の読み込み
```{r}
paste(curdir, "/", organ,"/Marker_gene/Hs_", organ,
      ".seurat_integrate.celltype_pick_markergene.tsv", sep = "") %>%
  read_tsv(show_col_types = FALSE) -> Hs_all.marker
paste(curdir, "/", organ,"/Marker_gene/Mm_", organ,
      ".seurat_integrate.celltype_pick_markergene.tsv", sep = "") %>%
  read_tsv(show_col_types = FALSE) -> Mm_all.marker
```

```{r}
human_table <- tibble()
mouse_talbe <- tibble()

# 1. クラスター選択
clstr <- 0
# 2. マーカー遺伝子抽出
Hs_all.marker %>%
  filter(cluster == clstr) -> Hs_all.marker.ident0
Mm_all.marker %>%
  filter(cluster == clstr) -> Mm_all.marker.ident0
# 3.4 表を結合 
human_mouse_ortholog %>%
  inner_join(Hs_all.marker.ident0, by = c("Gene_name" = "gene" )) %>%
  inner_join(Mm_all.marker.ident0, by = c("Mouse_gene_name" = "gene", "cluster")) %>%
  arrange(p_val_adj.x) %>%
  select(c(1:3)) %>%
  as_tibble -> Hs_Mm.shared_marker.ident0
# 共通遺伝子をデータフレームに保存
Hs_Mm.shared_marker.ident0 %>%
  inner_join(Hs_all.marker.ident0,
             by = c("Gene_name" = "gene")) %>%
  mutate(share = "shared") %>%
  bind_rows(human_table, .) -> human_table

```

もともとやろうとしていたこと
遺伝子対応関係表に対応する遺伝子を追加
```{r}
# ヒト->マウスの表
read_tsv("Gene_name/20230112_Hs_Mm_gene_name_release104.txt",
         show_col_types = FALSE) %>%
  mutate(Homology_type = replace(Mouse_homology_type,
                                 is.na(Mouse_homology_type),
                                 "SS")) %>%
  dplyr::select(c(Gene_name, Mouse_gene_name, Homology_type)) %>%
  distinct -> human_mouse_ortholog_raw

# マウス->ヒトの表
read_tsv("Gene_name/20230112_Mm_Hs_gene_name_release104.txt",
         show_col_types = FALSE) %>%
  mutate(Homology_type = replace(Human_homology_type,
                                　is.na(Human_homology_type),
                                  "SS")) %>%
  dplyr::select(c(Gene_name, Human_gene_name, Homology_type)) %>%
  distinct -> mouse_human_ortholog_raw

full_join(human_mouse_ortholog_raw,
          mouse_human_ortholog_raw,
          by = c("Gene_name" = "Human_gene_name",
                 "Mouse_gene_name" = "Gene_name")) %>%
  distinct %>%
  write.table("Gene_name/human-mouse_gene_name_full_join_table.tsv",
              sep = "\t", quote = F, row.names = F)
```

前に作成したOrthology表と結合させる
```{r}
#　これまでのヒト遺伝子対応表、マウス遺伝子対応表
read_tsv("Gene_name/Hs_Gene_name.homology_type.not_uniq_relation.manual_curate.txt",
         show_col_types = FALSE) -> human_orthology
read_tsv("Gene_name/Mm_Gene_name.homology_type.not_uniq_relation.manual_curate.txt",
         show_col_types = FALSE) -> mouse_orthology

# Seuratで使われている遺伝子名を取得
Mm_seurat@assays$RNA@counts %>%
  rownames %>%
  data.frame(Mouse_gene_name = .) %>%
  as_tibble -> Mm_seurat_gene
Hs_seurat@assays$RNA@counts %>%
  rownames %>%
  data.frame(Gene_name = .) %>%
  as_tibble -> Hs_seurat_gene

# 結合
human_mouse_ortholog_raw %>%
    inner_join(human_orthology,
               by = c("Gene_name",
                      "Homology_type" = "Mouse_homology_type.x")) %>%
    full_join(mouse_orthology,
              by = c("Mouse_gene_name" = "Gene_name")) %>%
    arrange(Gene_name) %>%
    mutate(Homology_type = replace(Homology_type,
                                   is.na(Homology_type),
                                   "SS"),
           Human_homology_type.x = replace(Human_homology_type.x,
                                   is.na(Human_homology_type.x),
                                   "SS")) %>%
    filter(Homology_type != Human_homology_type.x) %>%
    inner_join(Mm_seurat_gene, by = "Mouse_gene_name")
```
1/17: ヒト-マウスで登録されている対応関係が異なる遺伝子がある
1. マウスでは名前(Gene_name)がついているが、ヒトでは名前がついていない遺伝子がある。この場合もENSEMBL IDはある
  eg. Lhb
2. マウスではone-to-manyになっている遺伝子があった。これはアノテーションのミスかもしれない
  eg. HsPCDHA4にMmPcdha4とMmPcdha11が対応する遺伝子として登録されていた。ヒト->マウスはone-to-manyになっている。マウスでは異なるENSEMBL IDにPcdha11が登録されていた。うち片方はPcdha4と同じゲノム領域から抽出された遺伝子だった。
3. Tabula sapiensで解析に使われた遺伝子とENSEMBLに登録されている遺伝子名が異なる遺伝子があった
  Tabula sapiens:AL627309 -> ENSEMBL:OR4F5
  ただし、これはENSEMBLのGRCh37にはある遺伝子名だった。Tabula sapiensが利用したゲノムアノテーション情報にあたった方が良さそう
  -> GRCh38だが、Gencode Reference 30を利用している、おそらくtranscript
　-> transcriptの中に含まれていた。この情報をもとにすれば、Biomartの中で遺伝子名が含まれていなかった配列に名前をつけて取り扱うことができそう
　Tabula murisも確認した方が良さそう
　-> mm10plus genome (https://www.nature.com/articles/s41586-018-0590-4#Sec7)
　-> ensembl Biomartからsynonymを利用すれば解決する？



# ---1/20、結局使わなかった部分
```{r}
inner_join(MM10_PLUS, Mm_seurat_gene,
           by = c("Gene_name" = "Mouse_gene_name")) # 23,337 genes
Mm_seruat_gene # 23,341 genes

# MGI全対応関係の表だと理解
read_tsv("Gene_name/MGI/MRK_Sequence.rpt",
         show_col_types = FALSE) -> mrk_sequence

# Refseq IDが変わってたが、遺伝子名から復元できた
bind_rows(refseq_mm10_mrna, refseq_mm10_lnrna) %>%
  anti_join(Mm_seurat_gene, ., by = c("Mouse_gene_name" = "Gene_name.y")) %>%
  distinct() %>%
  inner_join(ensembl_refseq,
             by = c("Mouse_gene_name" = "Gene_name")) %>%
  select(c("Mouse_gene_name", "Gene_stable_ID")) %>%
  distinct -> refseq_mm10_rename # 360 genes, 361 gene ID
# シノニムが使われていた
bind_rows(refseq_mm10_mrna, refseq_mm10_lnrna) %>%
    anti_join(Mm_seurat_gene, ., by = c("Mouse_gene_name" = "Gene_name.y")) %>%
    distinct() %>%
    anti_join(ensembl_refseq,
              by = c("Mouse_gene_name" = "Gene_name")) %>%
    inner_join(Mm_synomym,
               by = c("Mouse_gene_name" = "Gene_Synonym")) %>%
  distinct -> refseq_mm10_synonym # 97 genes, 97 gene ID
# MGIには遺伝子名があり、ENSEMBL IDに変換できた 
bind_rows(refseq_mm10_mrna, refseq_mm10_lnrna) %>%
    anti_join(Mm_seurat_gene, ., by = c("Mouse_gene_name" = "Gene_name.y")) %>%
    distinct() %>%
    anti_join(ensembl_refseq,
              by = c("Mouse_gene_name" = "Gene_name")) %>%
    anti_join(Mm_synomym,
              by = c("Mouse_gene_name" = "Gene_Synonym")) %>%
  inner_join(mgi_ensembl,
             by = c("Mouse_gene_name" = "Marker_Symbol")) %>%
  select(c("Mouse_gene_name", "Gene_stable_ID")) %>%
  filter(!is.na(Gene_stable_ID)) -> refseq_mm10_mgi # 16 genes, 16 gene ID

bind_rows(refseq_mm10_mrna, refseq_mm10_lnrna) %>%
    anti_join(Mm_seurat_gene, ., by = c("Mouse_gene_name" = "Gene_name.y")) %>%
    distinct() %>%
    anti_join(ensembl_refseq,
              by = c("Mouse_gene_name" = "Gene_name")) %>%
    anti_join(Mm_synomym,
              by = c("Mouse_gene_name" = "Gene_Synonym")) %>%
    anti_join(mgi_ensembl,
              by = c("Mouse_gene_name" = "Marker_Symbol")) %>%
  inner_join(MM10_PLUS, by = c("Mouse_gene_name" = "Gene_name")) %>%
  inner_join(mrk_sequence, by = c("Refseq" = "RefSeq_transcript_IDs")) %>%
  select("Mouse_gene_name", "Ensembl_transcript_IDs") %>%
  filter(!is.na(Ensembl_transcript_IDs))
```
それでもなお、ENSEMBLにない遺伝子がある
0610010B08Rik -> MGI:3780195 -> ENSMUSG00000078886 -> Gm2026 (synonymなし)
つまり、MGIのIDに変換を挟めばENSEMBL IDに変換できそう

MGI IDに変換する
```{r}
# Tabula murisにあって、マウスベースのヒト-マウス対応表にない遺伝
Mm_seurat_gene %>%
  inner_join(Mm_ortholog_raw,
            by = c("Mouse_gene_name" = "Gene_name")) %>%
  dplyr::select(c("Mouse_gene_name", "Gene_stable_ID")) %>%
  distinct -> Mm_seurat_gene.in_ensembl # 20472genes
# Tabula murisにあって、マウスベースのヒト-マウス対応表にない遺伝子
Mm_seurat_gene %>%
  anti_join(Mm_ortholog_raw,
            by = c("Mouse_gene_name" = "Gene_name")) -> Mm_seurat_gene.no_biomart # 2896 genes

# Tabula murisにある遺伝子のうち、ENSEMBLのシノニム表にある遺伝子を取得
read_tsv("Gene_name/20230117_Mm_synonym_release104.txt",
         show_col_types = FALSE) -> Mm_synonym
Mm_seurat_gene.no_biomart %>%
  inner_join(Mm_synonym,
             by = c("Mouse_gene_name" = "Gene_Synonym" )) %>%
  dplyr::select(c("Mouse_gene_name", "Gene_stable_ID", "Gene_name")) %>%
  distinct -> Mm_seurat_gene.synonym # 2441 genes

# 上を使って、マウスベースのヒト-マウス対応表とENSEMBLのシノニムの両方で見つからない遺伝子
Mm_seurat_gene.no_biomart %>%
  anti_join(Mm_no_ensembl_gene_name,
            by = "Mouse_gene_name") -> Mm_seurat_gene.no_synonym # 472 genes


# List of Mouse Genetic Markers
read_tsv("Gene_name/MGI/MRK_List1.rpt",
         show_col_types = FALSE) -> mrk_list
# MGI Marker Coordinates
read_tsv("Gene_name/MGI/MGI_Gene_Model_Coord.rpt",
         show_col_types = FALSE) -> mgi_coord


# MGI全対応関係のうち、Tabula murisに含まれている遺伝子
Mm_seurat_gene %>%
  inner_join(mrk_sequence,
             by = c("Mouse_gene_name" = "Marker_Symbol")) -> Mm_seurat_gene.mrk_gene # 19,426 genes
# 上のうち、ENSEMBL Transcript IDが振られていないもの
Mm_seurat_gene.mrk_gene  %>%
  filter(is.na(Ensembl_transcript_IDs)) # 199 genes

# Tabula murisとMGI databaseのENSEMBL-Gene nameの対応関係表に共通する遺伝子
Mm_seurat_gene %>%
  inner_join(mgi_coord,
             by = c("Mouse_gene_name" = "marker_symbol")) %>% 
  select(c("Mouse_gene_name", "Ensembl_gene_id"))-> Mm_seurat_gene.gene_mgi_coord # 19,377遺伝子
# Tabula murisのうち、MGI databaseのENSEMBL-Gene nameの対応関係表にない遺伝子を取得
Mm_seurat_gene %>%
  anti_join(mgi_coord,
            by = c("Mouse_gene_name" = "marker_symbol")) -> Mm_seurat_gene.gene_no_mgi_coord # 3,954遺伝子
```
シノニムが見つからない遺伝子について、MGIで復元できないか？
```{r}
Mm_seurat_gene.no_synonym %>%
  anti_join(mgi_coord,
            by = c("Mouse_gene_name" = "marker_symbol")) -> Mm_seurat_gene.no_mgi_coord # 248 genes
# マーカー情報を参照してみる
Mm_seurat_gene.no_mgi_coord %>%
  anti_join(mrk_list, by = c("Mouse_gene_name" = "Marker_Symbol")) -> Mm_seurat_gene.no_mgi # 114 genes
# 遺伝子の名前の付け方が異なるものが混ざっている
Mm_seurat_gene.no_mgi %>%
  mutate(Mouse_gene_name2 = paste(Mouse_gene_name, "Rik", sep = "")) %>%
  anti_join(mgi_coord,
            by = c("Mouse_gene_name2" = "marker_symbol")) -> Mm_seurat_gene.no_relation # 97 genes
# ここまでくるとMGIにない、またはMGI IDを使ってもENSEMBL IDに変換できない。

# ここで変換できたMGI accession IDはENSEMBL IDに変換できるのか？
Mm_seurat_gene.no_synonym %>%
    inner_join(mgi_coord,
              by = c("Mouse_gene_name" = "marker_symbol")) %>%
  filter(!Ensembl_gene_id == "null") %>%
  select(c("Mouse_gene_name", "Ensembl_gene_id")) %>%
  dplyr::rename(Gene_stable_ID = Ensembl_gene_id ) -> Mm_seurat_gene.no_synonym.mgi_coord # 88遺伝子はENSEMBL IDに変換できた
# マーカー情報を参照したもの
Mm_seurat_gene.no_mgi_coord %>%
    inner_join(mrk_list,
               by = c("Mouse_gene_name" = "Marker_Symbol")) %>%
  select(c("MGI_Accession_ID", "Mouse_gene_name")) %>%
  filter(!MGI_Accession_ID == "NULL") %>%
  inner_join(mgi_coord, by = c("MGI_Accession_ID" = "MGI_accession_id")) # 0 gene, つまりなし
# 名前変更したら見つかったもの
Mm_seurat_gene.no_mgi %>%
    mutate(Mouse_gene_name2 = paste(Mouse_gene_name, "Rik", sep = "")) %>%
    inner_join(mgi_coord,
              by = c("Mouse_gene_name2" = "marker_symbol")) %>%
  filter(!Ensembl_gene_id == "null") %>% 
  select(c("Mouse_gene_name", "Ensembl_gene_id")) %>%
  dplyr::rename(Gene_stable_ID = Ensembl_gene_id ) -> Mm_seurat_gene.rename_mgi_coord
```
マウス統合
```{r}
full_join(Mm_ortholog_raw, mgi_coord,
           by = c("Gene_stable_ID" = "Ensembl_gene_id")) %>%
  dplyr::select(c("Gene_stable_ID", "Gene_name", "Human_gene_stable_ID",
                  "Human_homology_type")) %>%
  distinct() -> Mm_mgi_orthology
# filter(Gene_name == marker_symbol) 

bind_rows(Mm_seurat_gene.in_ensembl, Mm_seurat_gene.synonym,
          Mm_seurat_gene.no_synonym.mgi_coord,
          Mm_seurat_gene.rename_mgi_coord) %>%
  select(c("Mouse_gene_name", "Gene_stable_ID")) -> Mm_seurat_mgi # 23,017 genes / 23,341
Mm_mgi_orthology %>%
  mutate(Human_homology_type = replace(Human_homology_type,
                                       is.na(Human_homology_type),
                                       "SS")) %>%
  full_join(Mm_seurat_mgi, by = "Gene_stable_ID") %>%
  dplyr::select(c("Mouse_gene_name", "Gene_stable_ID", "Gene_name",
                  "Human_gene_stable_ID", "Human_homology_type")) -> Mm_mgi_seurat_ensembl
Mm_mgi_seurat_ensembl %>%
  write.table("Gene_name/20230119_Mm_Hs_genecode_muris.tsv",
              sep = "\t", quote = F, row.names = F)
```

gencode(G), tabula sapiens(T), ENSEMBL biomart(E)に共通する遺伝子名一覧を取得
数える用: %>% select("Gene_stable_ID", "Gene_name") %>% distinct
G: 58,825
T: 58,870
E: 67,128
```{r}
Hs_ortholog_raw %>%
  inner_join(Hs_gencode,
            by = c("Gene_stable_ID" = "ENSEMBL_gene_ID")) %>%
  select(c("Gene_stable_ID", "Gene_name.x", "Gene_name.y",
           "Mouse_gene_stable_ID", "Mouse_homology_type")) %>%
  distinct -> Hs_g_t
# G & T & E
Hs_ortholog_raw %>%
  inner_join(Hs_gencode,
            by = c("Gene_stable_ID" = "ENSEMBL_gene_ID")) %>%
  select(c("Gene_stable_ID", "Gene_name.x", "Gene_name.y",
           "Mouse_gene_stable_ID", "Mouse_homology_type")) %>%
  distinct %>%
  filter(Gene_name.y == Gene_name.x) %>%
  inner_join(Hs_seurat_gene, .,
            by = c("Gene_name" = "Gene_name.y")) %>%
  select("Gene_stable_ID") %>%
  distinct # 37,028

# Gene nameをgencodeに合わせると見つかるもの
Hs_ortholog_raw %>%
    inner_join(Hs_gencode,
               by = c("Gene_stable_ID" = "ENSEMBL_gene_ID")) %>%
    select(c("Gene_stable_ID", "Gene_name.x", "Gene_name.y",
             "Mouse_gene_stable_ID", "Mouse_homology_type")) %>%
    distinct %>%
    filter(Gene_name.y != Gene_name.x) %>%
    inner_join(Hs_seurat_gene, .,
               by = c("Gene_name" = "Gene_name.y")) %>%
    select("Gene_stable_ID") %>%
    distinct # 3513

# Gene nameをENSEMBL biomartに合わせると見つかるもの
Hs_ortholog_raw %>%
    inner_join(Hs_gencode,
               by = c("Gene_stable_ID" = "ENSEMBL_gene_ID")) %>%
    select(c("Gene_stable_ID", "Gene_name.x", "Gene_name.y",
             "Mouse_gene_stable_ID", "Mouse_homology_type")) %>%
    distinct %>%
    filter(Gene_name.y != Gene_name.x) %>%
    inner_join(Hs_seurat_gene, .,
               by = c("Gene_name" = "Gene_name.y")) %>%
    select("Gene_stable_ID") %>%
  distinct # 91

# !G & !E & T
anti_join(Hs_seurat_gene, Hs_g_t,
          by = c("Gene_name" = "Gene_name.x")) %>%
  anti_join(Hs_g_t, by = c("Gene_name" = "Gene_name.y")) # 1,767

# G & E! & T
Hs_g_t %>%
  filter(is.na(Mouse_homology_type)) %>%
  inner_join(Hs_seurat_gene, .,
             by = c("Gene_name" = "Gene_name.y")) %>%
    select("Gene_stable_ID") %>%
  distinct# 221
# G & E! & !T
Hs_g_t %>%
  filter(is.na(Mouse_homology_type)) %>%
  anti_join(Hs_seurat_gene,
             by = c("Gene_name.y" = "Gene_name")) # 0

# !G & E & T
Hs_g_t %>%
  filter(is.na(Gene_name.y)) %>%
  inner_join(Hs_seurat_gene,
             by = c("Gene_name.x" = "Gene_name")) %>%
    select("Gene_stable_ID") %>%
  distinct # 4,405

# !G & E & !T
Hs_g_t %>%
  filter(is.na(Gene_name.y)) %>%
  anti_join(Hs_seurat_gene,
             by = c("Gene_name.x" = "Gene_name")) %>%
    select("Gene_stable_ID") %>%
  distinct # 4,119


```



